src/MLIAP/mliap_so3_math.h

```diff
@@ -1,684 +1,181 @@
+int invert_matrix( int n,double *A, double *Ainv );
+int LUPdecompose( int n, double dtol, double *A, int *P );
+void LUPSolve( int n,double *A, double *B, int *P);
 
-/// @brief  Allocate a 2-dimensional array.  (Uses row-major order.)
-template<typename Entry>
-void Alloc2D(size_t nrows, size_t ncols, Entry ***paaX);
-
-/// @brief  Deallocate arrays that were created using Alloc2D().
-template<typename Entry>
-void Dealloc2D(Entry ***paaX);
-
-// ---- implementation ----
-
-template<typename Entry>
-void Alloc2D(size_t nrows, size_t ncols, Entry ***paaX)
-{
-  assert(paaX);
-  *paaX = new Entry* [nrows];
-  (*paaX)[0] = new Entry [nrows * ncols];
-  for(size_t iy=0; iy<nrows; iy++)
-    (*paaX)[iy] = (*paaX)[0] + iy*ncols;
-  // The caller can access the contents using (*paaX)[i][j]
-}
-
-template<typename Entry>
-void Dealloc2D(Entry ***paaX)
-{
-  if (paaX && *paaX) {
-    delete [] (*paaX)[0];
-    delete [] (*paaX);
-    *paaX = nullptr;
-  }
 }
 
-} // namespace matrix_alloc_jpd
-
-#endif //#ifndef _MATRIX_ALLOC_JPD_H
-
-
-#ifndef _JACOBI_HPP
-#define _JACOBI_HPP
-
-#include <algorithm>
-#include <cmath>
+using namespace MathEigen;
 
-
-namespace jacobi_pd {
-
-using namespace matrix_alloc_jpd;
-
-
-template<typename Scalar,
-         typename Vector,
-         typename Matrix,
-         typename ConstMatrix=Matrix>
-class Jacobi
+typedef Jacobi<double,double*,double**,double const*const*> Jacobi_v2;
+int SO3Math::jacobin(int n, double const* const* mat, double *eval, double **evec)
 {
-  int n;
-  Scalar **M;
-  Scalar c;
-  Scalar s;
-  Scalar t;
-  int *max_idx_row;
-
- public:
-
-  void SetSize(int n);
-
-  Jacobi(int n = 0) {
-    Init();
-    SetSize(n);
-  }
-
-  ~Jacobi() {
-    Dealloc();
-  }
-
-  typedef enum eSortCriteria {
-    DO_NOT_SORT,
-    SORT_DECREASING_EVALS,
-    SORT_INCREASING_EVALS,
-    SORT_DECREASING_ABS_EVALS,
-    SORT_INCREASING_ABS_EVALS
-  } SortCriteria;
-
-  int Diagonalize(ConstMatrix mat, Vector eval, Matrix evec,
-                  SortCriteria sort_criteria=SORT_DECREASING_EVALS,
-                  bool calc_evecs=true, int max_num_sweeps = 50);
-
-  int Diagonalize_1DArray(double *mat, Vector eval, double *evec,
-                          SortCriteria sort_criteria=
-                          SORT_DECREASING_EVALS,
-                          bool calc_evecs=true,
-                          int max_num_sweeps = 50);
-
- private:
-
-  void CalcRot(Scalar const *const *M, int i, int j);
-
-  void ApplyRot(Scalar **M, int i, int j);
-
-  void ApplyRotLeft(Matrix E, int i, int j);
-
-  void ApplyRotLeft_1DArray(double *E, int i, int j);
-
-  int MaxEntryRow(Scalar const *const *M, int i) const;
-
-  void MaxEntry(Scalar const *const *M, int& i_max, int& j_max) const;
-
-  void SortRows(Vector v, Matrix M, int n,
-                SortCriteria s=SORT_DECREASING_EVALS ) const;
-
-  void Alloc(int N);
-  void Init();
-  void Dealloc();
+  int i,j;
+  double mat_cpy[n][n];
```

Comment:
This statement is not legal C++ it is only acceptable in C98 (and allowing it is considered a mistake). GCC will warn about it with -Wvla. You have to either explicitly allocate and delete it or use std::unique_ptr.